datashader/bokeh_ext.py

```diff
@@ -188,18 +208,35 @@ def update(self, ranges):
             Dictionary with of x/y-ranges, width and height.
         """
         if not self.comms_handle:
-            self.comms_handle = _CommsHandle(get_comms(self.ref),
-                                             self.doc, {})
+            comm = get_comms(self.ref)
+            comm_args = (comm, self.doc) if bokeh_version > '0.12.9' else (comm, self.doc, {})
+            self.comms_handle = CommsHandle(*comm_args)
 
         self.update_image(ranges)
         msg = self.get_update_event()
-        self.comms_handle.comms.send(msg)
+        comm = self.comms_handle.comms
+        if bokeh_version > '0.12.9':
+            comm.send(msg.header_json)
+            comm.send(msg.metadata_json)
+            comm.send(msg.content_json)
+            for header, payload in msg.buffers:
+                comm.send(json.dumps(header))
+                comm.send(buffers=[payload])
+        else:
+            comm.send(msg)
         return 'Complete'
 
     def get_update_event(self):
         """
         Generate an update event json message.
         """
+        if bokeh_version > '0.12.9':
+            events = list(self.doc._held_events)
+            if not events:
+                return None
+            msg = Protocol("1.0").create("PATCH-DOC", events)
+            self.doc._held_events = []
+            return msg
```

Comment:
The 0.12.9 and previous versions are so entirely different that again I think it would be helpful if you wrapped up what each one does as some sort of coherent function, with a reasonable name, and then simply switched between them.  I'd rather such complexities be put into separate functions that can be ignored when understanding the overall logic of this class.